### DISCUSSION TITLE
is posible read and get a user in all tree and branch of an active directory ?

### DISCUSSION THREAD

**Discussion by hdkcreative** [DirectoryTree/LdapRecord#631] [General]
Hi guys, a question, this library has the option of read  an user in all tree and branch only with your user and your  password ?

example: this is our AD
![imagen](https://github.com/DirectoryTree/LdapRecord/assets/18191896/fae5816a-743b-4438-b3c0-5a9c6cb7dd04)

COORP-GROUP.com is our principal entity ( corporate)

mexico, canda, japon , spain and USA are our branch of the principal AD

is posible only with a form with   username ( samaAccountname) , domain and password serach and locate a user in our AD ?

with this library we can read in cascade all ad of the principal ?

**Comment by hdkcreative** [comment]
**first approach :** 

We did it this way to go through all the Active Directory and search for a user, we also did `echo date('H:m:s')`  to measure the response time (maximum 2 seconds so far), we have 300 users in total among all active directories

```
$username = 'Jhoe.Doe';
echo 'Search Started: '.date('H:m:s').'<br>';

            $activeDirectories = ['mexico', 'canada', 'japon' , 'spain' , 'USA' ];

            foreach ($activeDirectories as $directory) {
                $user = LdapUser::on($directory)->where('samaccountname', $username)->first();
                
                if (!$user) {
                    var_dump(" the User ".$username." Not  EXISTs : in $directory<br>");
                } else {
                    var_dump(" the User ".$username." EXISTs in $directory<br>");
                }
            }

            echo 'Search Finish in :  '.date('H:m:s').'<br>';
            exit;
```

We're still waiting to see if anyone knows of a recursive way to traverse active directory to find a user.

**Reply by stevebauman** [reply]
Hi @hdkcreative,

LdapRecord uses PHP's `ldap_search` method by default when performing LDAP queries, which is recursive search.

What is your `base_dn` set to? By the looks of it, it should be set to `dc=coorp-group.com` to be able to perform searches in all DC's below it.

**Reply by hdkcreative** [reply]
Hi @stevebauman , thank you for your response :)

yes our base_dn is :  dc=coorp-group,dc=com

Is there somewhere in the documentation how to do that recursive search in the entire active directory?

we would need to test with php code

**Reply by stevebauman** [reply]
Happy to help @hdkcreative.

> Is there somewhere in the documentation how to do that recursive search in the entire active directory?

All searches are recursive by default. If your `base_dn` in your configured connection is `dc=coorp-group,dc=com`, then all searches performed using all models will use that as the search base:

```php
use LdapRecord\Container;
use LdapRecord\Connection;
use LdapRecord\Models\Entry;

$connection = new Connection([
    // ...
    'base_dn' => 'dc=coorp-group,dc=com',
]);

Container::addConnection($connection);

Entry::get(); // Retrieves all objects beneath "dc=coorp-group,dc=com".
```

**Reply by hdkcreative** [reply]
Thank you @stevebauman , now i will to try check this and do test

**Reply by hdkcreative** [reply]
**it's  Work !** 👍  :)

thank you @stevebauman 

**our code :**


```
try {
            $connection = new Connection([
                'hosts' => ['40.9*9.1*4.7'],
                'username' => 'Administrator@coorp-group.com',
                'password' => '***1234****',
                'base_dn' => 'dc=coorp-group,dc=com',
                'port' =>  389,
            ]);
            $connection->connect();
            
            Container::addConnection($connection);
            $response =  Entry::get();
            var_dump($response [0]["subrefs"]); 
        
            echo "Successfully connected!";
        } catch (\LdapRecord\Auth\BindException $e) {
            $error = $e->getDetailedError();
        
            echo $error->getErrorCode();
            echo $error->getErrorMessage();
            echo $error->getDiagnosticMessage();
        }


var_dump output

array(8) {
[0]=>
string(35) "DC=canada,DC=coorp-group,DC=com"
[1]=>
string(38) "DC=usa,DC=coorp-group,DC=com"
[2]=>
string(35) "DC=japan,DC=coorp-group,DC=com"
[3]=>
string(36) "DC=spain,DC=coorp-group,DC=com"
[4]=>
string(34) "DC=mexico,DC=coorp-group,DC=com"
[5]=>
string(43) "DC=ForestDnsZones,DC=coorp-group,DC=com"
[6]=>
string(43) "DC=DomainDnsZones,DC=coorp-group,DC=com"
[7]=>
string(42) "CN=Configuration,DC=coorp-group,DC=com"
}


```

**Reply by stevebauman** [reply]
Excellent! Glad you're up and running @hdkcreative 👏
